blockchain/client.go

```diff
@@ -95,50 +191,101 @@ func (c *Client) StartEventsListening() (func(), <-chan error, error) {
 						continue
 					}
 
-					for _, callback := range c.eventsListeners {
-						go callback(events, header.Number, set.Block)
+					if cancelled {
+						return
 					}
+
+					callback(events, header.Number, set.Block)
 				}
 			}
 		}
 	}()
 
 	once := sync.Once{}
-	c.stopListening = func() {
+	cancel := func() {
 		once.Do(func() {
-			done <- struct{}{}
-			sub.Unsubscribe()
-			c.isListening = 0
+			c.mu.Lock()
+			cancelled = true
+			delete(c.eventsListeners, idx)
+			c.mu.Unlock()
 		})
 	}
 
-	return c.stopListening, c.errsListening, nil
+	return cancel, nil
 }
 
-func (c *Client) RegisterEventsListener(callback EventsListener) (func(), error) {
-	var idx int
-	for i := 0; i <= math.MaxInt; i++ {
-		if _, ok := c.eventsListeners[i]; !ok {
-			idx = i
-			break
+func (c *Client) processSystemEventsStorageChanges(
```

Comment:
what is system events storage changes?

Reply:
Blockchain node API has an RPC method to subscribe to the database key changes. In that function `System` is a pallet, `Events` is the storage item name which together identify the key-value pair. Storage changes is a type which subscription method returns to us. And this function is a handler which decodes value changes to system events. Replaced with less specific `onChanges` function which works for historical blocks events processing too.